### DISCUSSION TITLE
Use full window height

### DISCUSSION THREAD

**Discussion by jammerhund** [zauberzeug/nicegui#4049] [Q&A]
### Question

Hi,

i try to build a simple layout with a mainarea and a small statearea on bottom of the browser window. The mainarea should use the full height of the window. But it does only use the height of the  ui.label('Mainarea'). How can i do this?

```py
from nicegui import ui

with ui.row().classes('border full-width full-height'):
    ui.label('Mainarea')

with ui.row().classes('border full-width'):
    ui.label('Statearea')

ui.run()
```

**Comment by rodja** [comment]
Do you know about `ui.footer`?
https://nicegui.io/documentation/page_layout

**Comment by falkoschindler** [comment]
Hi @jammerhund,

"Full width" and "full height" usually refer to the size of the parent container. In case of your rows, there is no container except of the web page itself. And since a web page has basically infinite (or undefined) height, setting an element's height to "full" (or 100%) has no effect. But we can introduce a container, like the `ui.column` in the following example, with a specific height of `calc(100vh-2rem)` This starts with `100vh`, with "[vh](https://www.w3schools.com/cssref/css_units.php)" being 1% of the height of the view port, and subtracts 2rem to compensate for the page padding:

```py
with ui.column().classes('border w-full h-[calc(100vh-2rem)]'):
    with ui.row().classes('border w-full h-full'):
        ui.label('main area')
    with ui.row().classes('border w-full'):
        ui.label('state area')
```

Last but not least, letting the main area grow to full height automatically leaves enough space for the state area.

By the way, I'm using Tailwind classes "w-\*" and "h-\*" here for their flexibility when it comes to custom values.

See https://github.com/zauberzeug/nicegui/discussions/3079#discussioncomment-9454323 for a similar discussion.

**Reply by v479038280** [reply]
```py
ui.context.client.page_container.default_slot.children[0].props(
    ''':style-fn="(offset, height) => ( { height: offset ? `calc(100vh - ${offset}px)` : '100vh' })"'''
)
ui.context.client.content.classes("h-full")
```
if this one is better

**Reply by falkoschindler** [reply]
Great solution, @v479038280! I haven't thought of this approach.
This solution feels a bit more complicated, but works better than mine. If you, e.g., add a `ui.header`, the style fucntion still yields the correct page height. In my example, on the other hand, you would need to adjust the "2rem" to compensate the additional header.

**Reply by falkoschindler** [reply]
Slightly more compact:
```py
ui.context.client.page_container.default_slot.children[0].props(':style-fn="o => ({ height: `calc(100vh - ${o}px)` })"')
ui.context.client.content.classes('h-full')
```

Documentation for `style-fn`: https://quasar.dev/layout/page#style-fn

**Reply by MicaelJarniac** [reply]
> Slightly more compact:
> 
> ```python
> ui.context.client.page_container.default_slot.children[0].props(':style-fn="o => ({ height: `calc(100vh - ${o}px)` })"')
> ui.context.client.content.classes('h-full')
> ```
> 
> Documentation for `style-fn`: https://quasar.dev/layout/page#style-fn

This worked perfectly for me so far! I think it should at least be listed on the docs, or maybe even added as a built-in method one can call to automatically apply this.
